src/lib.rs

```diff
@@ -77,85 +79,324 @@
 //!# Ok(())
 //!# }
 //! ```
-//!
-mod ffi;
-#[macro_use]
-mod macros;
 
-fd_guard!(StdoutOverride, guard: StdoutOverrideGuard, FD: crate::ffi::STDOUT_FILENO, name of FD: stdout);
-fd_guard!(StdinOverride, guard: StdinOverrideGuard, FD: crate::ffi::STDIN_FILENO, name of FD: stdin);
-fd_guard!(StderrOverride, guard: StderrOverrideGuard, FD: crate::ffi::STDERR_FILENO, name of FD: stderr);
+use std::fs::File;
+use std::io::{self, IoSlice, IoSliceMut, Read, Write};
+use std::mem::ManuallyDrop;
+use std::path::Path;
+
+#[cfg(not(any(unix)))]
+compile_error!("stdio-override only supports Unix");
+
+#[cfg_attr(unix, path = "unix.rs")]
+mod imp;
+
+/// An overridden standard input.
+///
+/// Reading from this reads the original standard input. When it is dropped the standard input
+/// will be reset.
+#[derive(Debug)]
+pub struct StdinOverride {
+    original: ManuallyDrop<File>,
+    reset: bool,
+}
+impl StdinOverride {
+    /// Read standard input from the raw file descriptor. The file descriptor must be readable.
+    ///
+    /// The file descriptor is not owned, so it is your job to close it later. Closing it while
+    /// this exists will not close the standard error.
+    pub fn from_raw(raw: imp::Raw) -> io::Result<Self> {
+        Ok(Self { original: ManuallyDrop::new(imp::override_stdin(raw, false)?), reset: false })
+    }
+    /// Read standard input from the owned raw file descriptor. The file descriptor must be
+    /// readable.
+    ///
+    /// The file descriptor is owned, and so you must not use it after passing it to this function.
+    pub fn from_raw_owned(raw: imp::Raw) -> io::Result<Self> {
+        Ok(Self { original: ManuallyDrop::new(imp::override_stdin(raw, true)?), reset: false })
+    }
+    /// Read standard input from the IO device. The device must be readable.
+    ///
+    /// Dropping the IO device after calling this function will not close the standard input.
+    pub fn from_io_ref<T: imp::AsRaw>(io: &T) -> io::Result<Self> {
+        Self::from_raw(imp::as_raw(io))
+    }
+    /// Read standard input from the IO device. The device must be readable.
+    pub fn from_io<T: imp::IntoRaw>(io: T) -> io::Result<Self> {
+        Self::from_raw_owned(imp::into_raw(io))
+    }
+    /// Read standard input from the file at that file path.
+    ///
+    /// The file must exist and be readable.
+    pub fn from_file<P: AsRef<Path>>(path: P) -> io::Result<Self> {
+        Self::from_io(File::open(path)?)
+    }
+    /// Reset the standard input to its state before this type was constructed.
+    ///
+    /// This can be called to manually handle errors produced by the destructor.
+    pub fn reset(mut self) -> io::Result<()> {
+        self.reset_inner()?;
+        self.reset = true;
+        Ok(())
```

Comment:
Why not just drop / call `reset_inner()` + `mem::forget` and then we won't need the `reset` bool?

Reply:
Oh yeah, there's nothing else in the type that could leak memory.